clash-cores/src/Clash/Cores/Xilinx/Ethernet/Gmii.hs

```diff
@@ -0,0 +1,136 @@
+  "LogiCORE IP Ethernet 1000BASE-X PCS/PMA or SGMII"
+
+-}
+module Clash.Cores.Xilinx.Ethernet.Gmii where
+
+import Clash.Explicit.Prelude
+
+import Clash.Annotations.Primitive
+import Clash.Cores.Xilinx.Ethernet.Gmii.BlackBoxes
+import Clash.Cores.Xilinx.Ethernet.Gmii.Types
+import Clash.Explicit.Prelude
+import Clash.Signal.Internal(ClockN, DiffClock(..))
+
+-- | Wrapper for the LogiCORE IP Ethernet 1000BASE-X PCS/PMA or SGMII.
+-- Configured to function as a GMII to SGMII bridge.
+gmiiSgmiiBridge ::
+  forall asyncDom sgmii625 gmiiRx125 gmiiTx125 .
+  ( KnownDomain asyncDom
+  , KnownDomain sgmii625
+  , KnownDomain gmiiRx125
+  , KnownDomain gmiiTx125
+  , DomainPeriod sgmii625 ~  Picoseconds 1600
+  , DomainPeriod gmiiRx125 ~ Nanoseconds 8
+  , DomainPeriod gmiiTx125 ~ Nanoseconds 8
+  , DomainActiveEdge sgmii625 ~ Rising
+  , DomainActiveEdge gmiiRx125 ~ Rising
+  , DomainActiveEdge gmiiTx125 ~ Rising
+  ) =>
+  -- | Reference clock coming from the Phy.
+  DiffClock sgmii625 ->
+  -- | Asynchronous reset for the bridge.
+  Reset asyncDom ->
+  -- | Link speed configuration.
+  Signal gmiiRx125 LinkSpeed ->
+  -- | Signal detect from the Phy.
+  Signal asyncDom SignalDetect ->
+  -- | Configuration for the bridge.
+  Signal gmiiRx125 Config ->
+  -- | Auto negotiation configuration for the bridge.
+  Signal gmiiRx125 AutoNegConfig ->
+  -- | Restart auto negotiation.
+  Signal gmiiRx125 RestartAutoNeg ->
+  -- | LVDS input from the Phy.
+  Signal sgmii625 Lvds ->
+  -- | GMII input from the MAC.
+  Signal gmiiTx125 Gmii ->
+  -- | Output record, see 'BridgeOutput'.
+  BridgeOutput gmiiRx125 sgmii625
+gmiiSgmiiBridge refClk refRst cLinkSpeed signalDetect bridgeConfig anConfig anRestart lvdsIn gmiiTx = BridgeOutput{..}
+ where
+  ( lvdsOutP
+    , lvdsOutN
+    , gmiiRxD
+    , gmiiRxDv
+    , gmiiRxEr
+    , bridgeClk125
+    , activeHighRxRst
+    , fmap fromStatusVector -> bridgeStatusVector
+    ) = gmiiSgmiiBridgeBb clockP clockN activeHighRefRst (getSpeed100 <$> cLinkSpeed)
+      (getSpeed10_100 <$> cLinkSpeed) signalDetect (toConfigVector <$> bridgeConfig)
+      anConfig anRestart lvdsInP lvdsInN gmiiTxD gmiiTxEn gmiiTxEr
+
+  activeHighRefRst = unsafeToActiveHigh refRst
+  bridgeRst125 = unsafeFromActiveHigh activeHighRxRst
+  bridgeLvdsOut = Lvds <$> lvdsOutP <*> lvdsOutN
+  (DiffClock clockP clockN) = refClk
+  (lvdsInP, lvdsInN) = (fmap lvdsP lvdsIn, fmap lvdsN lvdsIn)
+  (gmiiTxD, gmiiTxEn, gmiiTxEr) = unbundle $ fmap (\Gmii{..} -> (gmiiData, gmiiValid, gmiiError)) gmiiTx
+  bridgeGmiiRx = Gmii <$> gmiiRxD <*> gmiiRxDv <*> gmiiRxEr
+
+-- | Black box for the GMII to SGMII bridge.
+gmiiSgmiiBridgeBb ::
+  -- | P channel of the reference clock coming from the Phy.
+  Clock sgmii625 ->
+  -- | N channel of the reference clock coming from the Phy.
+  ClockN sgmii625 ->
+  -- | Asynchronous active high reset for the bridge.
+  Signal asyncDom Bool->
+  -- | Whether the link should be configured for 100 Mbit/s.
```

Comment:
Since this function is exported, the illegal settings should at least be documented, but preferably checked.